File: Clases 2018/Clases-C#/ClasePractica10_2018 (update tabla Bugs)/C#_TPS_BugTracker/TPS_BugTracker/GUILayer/Usuarios/frmABMUsuario.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using TPS_BugTracker.BusinessLayer;
using TPS_BugTracker.BusinessLayer.Services;

namespace TPS_BugTracker.GUILayer.Usuarios
{
    public partial class frmABMUsuario : Form
    {
        public frmABMUsuario()
        {
            InitializeComponent();
        }

        public enum Opcion
        {
            insert,
            update,
            delete
        }

        private Opcion _action = Opcion.insert;
        private UsuarioService oUsuarioService = new UsuarioService();
        private Usuario oUsuarioSelected;

        private void frmABMUsuario_Load(System.Object sender, System.EventArgs e)
        {
            llenarCombo(cbo_perfil, BDHelper.getBDHelper().ConsultaSQL("SELECT * From Perfiles WHERE id_perfil <> 1"), "n_perfil", "id_perfil");
            switch (_action)
            {
                case Opcion.insert:
                    {
                        this.Text = "Nuevo Usuario";
                        break;
                    }

                case Opcion.update:
                    {
                        this.Text = "Actualizar Usuario";
                        // Recuperar usuario seleccionado en la grilla 
                        mostrar_datos();
                        txt_nombre.Enabled = true;
                        txt_email.Enabled = true;
                        txt_email.Enabled = true;
                        txt_password.Enabled = true;
                        txt_confirmar_pass.Enabled = true;
                        cbo_perfil.Enabled = true;
                        break;
                    }

                case Opcion.delete:
                    {
                        mostrar_datos();
                        this.Text = "Habilitar/Deshabilitar Usuario";
                        txt_nombre.Enabled = false;
                        txt_email.Enabled = false;
                        txt_email.Enabled = false;
                        txt_password.Enabled = false;
                        cbo_perfil.Enabled = false;
                        txt_confirmar_pass.Enabled = false;
                        break;
                    }
            }
        }

        private void btn_salir_Click(System.Object sender, System.EventArgs e)
        {
            this.Dispose();
        }


        public void seleccionar_usuario(Opcion op, Usuario usuarioSelected)
        {
            _action = op;
            oUsuarioSelected = usuarioSelected;
        }

        private void mostrar_datos()
        {
            if (oUsuarioSelected != null)
            {
                txt_nombre.Text = oUsuarioSelected.nombre;
                txt_email.Text = oUsuarioSelected.email;
                txt_password.Text = oUsuarioSelected.password;
                txt_confirmar_pass.Text = txt_password.Text;
                cbo_perfil.Text = oUsuarioSelected.perfil;
            }
        }

        private void btn_aceptar_Click(System.Object sender, System.EventArgs e)
        {
            string str_sql = "";
            Usuario oUsuario;

            switch (_action)
            {
                case Opcion.insert:
                    {
                        if (existe_nombre() == false)
                        {
                            if (validar_campos())
                            {
                                oUsuario = new Usuario();
                                oUsuario.nombre = txt_nombre.Text;
                                oUsuario.password = txt_password.Text;
                                oUsuario.email = txt_email.Text;
                                oUsuario.id_perfil = cbo_perfil.SelectedValue.ToString();
                                oUsuario.estado = "S";

                                if (oUsuarioService.crearUsuario(oUsuario))
                                {
                                    MessageBox.Show("Usuario insertado!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information);
                                    this.Close();
                                }
                            }
                        }
                        else
                            MessageBox.Show("Nombre de usuario encontrado!. Ingrese un nombre diferente", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information);
                        break;
                    }

                case Opcion.update:
                    {
                        if (validar_campos())
                        {
                            oUsuarioSelected.nombre = txt_nombre.Text;
                            oUsuarioSelected.password = txt_password.Text;
                            oUsuarioSelected.email = txt_email.Text;
                            oUsuarioSelected.id_perfil = cbo_perfil.SelectedValue.ToString();

                            if (oUsuarioService.actualizarUsuario(oUsuarioSelected))
                            {
                                MessageBox.Show("Usuario actualizado!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information);
                                this.Dispose();
                            }
                            else
                                MessageBox.Show("Error al actualizar el usuario!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information);
                        }

                        break;
                    }

                case Opcion.delete:
                    {
                        if (MessageBox.Show("Seguro que desea habilitar/deshabilitar el usuario seleccionado?", "Aviso", MessageBoxButtons.OKCancel, MessageBoxIcon.Question) == DialogResult.OK)
                        {
                            if (oUsuarioSelected.estado == "N")
                                oUsuarioSelected.estado = "S";
                            else
                                oUsuarioSelected.estado = "N";

                            if (oUsuarioService.modificarEstadoUsuario(oUsuarioSelected))
                            {
                                MessageBox.Show("Usuario Habilitado/Deshabilitado!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information);
                                this.Dispose();
                            }
                            else
                                MessageBox.Show("Error al actualizar el usuario!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information);
                        }

                        break;
                    }
            }
        }

        private bool validar_campos()
        {
            // campos obligatorios
            if (txt_nombre.Text == string.Empty)
            {
                txt_nombre.BackColor = Color.Red;
                txt_nombre.Focus();
                return false;
            }
            else
                txt_nombre.BackColor = Color.White;

            if (txt_password.Text == string.Empty)
            {
                txt_password.BackColor = Color.Red;
                txt_password.Focus();
                return false;
            }
            else
                txt_password.BackColor = Color.White;

            if (txt_confirmar_pass.Text == string.Empty)
            {
                txt_confirmar_pass.BackColor = Color.Red;
                txt_confirmar_pass.Focus();
                return false;
            }
            else
                txt_confirmar_pass.BackColor = Color.White;

            if (cbo_perfil.Text == string.Empty)
            {
                cbo_perfil.BackColor = Color.Red;
                cbo_perfil.Focus();
                return false;
            }
            else
                cbo_perfil.BackColor = Color.White;

            if (txt_confirmar_pass.Text != txt_password.Text)
            {
                txt_confirmar_pass.BackColor = Color.Red;
                txt_password.BackColor = Color.Red;
                txt_confirmar_pass.Focus();
                return false;
            }
            else
            {
                txt_confirmar_pass.BackColor = Color.White;
                txt_password.BackColor = Color.White;
            }

            return true;
        }

        private bool existe_nombre()
        {
            return (oUsuarioService.validarNombreUsuario(txt_nombre.Text) != null);
        }


        private void llenarCombo(ComboBox cbo, Object source, string display, String value)
        {
            cbo.DataSource = source;
            cbo.DisplayMember = display;
            cbo.ValueMember = value;
            cbo.SelectedIndex = -1;
        }

        private void btn_cancelar_Click(object sender, EventArgs e)
        {
            this.Close();
        }
    }
}

File: Clases 2018/Clases-C#/Simulacro_parcial_2_2018 (creacion de bug e historico en transaccion)/VB_TPS_BugTracker/TPS_BugTracker/GUILayer/Usuarios/frmABMUsuario.vb
Partial Public Class frmABMUsuario
    Inherits Form
    Public Enum Opcion
        insert
        update
        delete
    End Enum

    Private _action As Opcion = Opcion.insert
    Private oUsuarioService As UsuarioService = New UsuarioService()
    Private oPerfilService As PerfilService = New PerfilService()
    Private oUsuarioSelected As Usuario

    Private Sub frmABMUsuario_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        GUIHelper.getHelper().llenarCombo(cbo_perfil, oPerfilService.consultarPerfiles(), "nombre", "id_perfil")

        Select Case _action
            Case Opcion.insert
                Me.Text = "Nuevo Usuario"
                Exit Select
            Case Opcion.update
                Me.Text = "Actualizar Usuario"
                mostrar_datos()
                txt_nombre.Enabled = True
                txt_email.Enabled = True
                txt_email.Enabled = True
                txt_password.Enabled = True
                txt_confirmar_pass.Enabled = True
                cbo_perfil.Enabled = True
                Exit Select
            Case Opcion.delete
                mostrar_datos()
                Me.Text = "Habilitar/Deshabilitar Usuario"
                txt_nombre.Enabled = False
                txt_email.Enabled = False
                txt_email.Enabled = False
                txt_password.Enabled = False
                cbo_perfil.Enabled = False
                txt_confirmar_pass.Enabled = False
                Exit Select
        End Select
    End Sub

    Public Sub seleccionar_usuario(ByVal op As Opcion, ByVal usuarioSelected As Usuario)
        _action = op
        oUsuarioSelected = usuarioSelected
    End Sub

    Private Sub mostrar_datos()
        If oUsuarioSelected IsNot Nothing Then
            txt_nombre.Text = oUsuarioSelected.nombre
            txt_email.Text = oUsuarioSelected.email
            txt_password.Text = oUsuarioSelected.password
            txt_confirmar_pass.Text = txt_password.Text
            cbo_perfil.Text = oUsuarioSelected.perfil
        End If
    End Sub

    Private Sub btn_aceptar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btn_aceptar.Click
        Dim str_sql As String = ""
        Dim oUsuario As Usuario

        Select Case _action
            Case Opcion.insert

                If existe_nombre() = False Then

                    If validar_campos() Then
                        oUsuario = New Usuario()
                        oUsuario.nombre = txt_nombre.Text
                        oUsuario.password = txt_password.Text
                        oUsuario.email = txt_email.Text
                        oUsuario.id_perfil = cbo_perfil.SelectedValue.ToString()
                        oUsuario.estado = "S"

                        If oUsuarioService.crearUsuario(oUsuario) Then
                            MessageBox.Show("Usuario insertado!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information)
                            Me.Close()
                        End If
                    End If
                Else
                    MessageBox.Show("Nombre de usuario encontrado!. Ingrese un nombre diferente", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information)
                End If

                Exit Select
            Case Opcion.update

                If validar_campos() Then
                    oUsuarioSelected.nombre = txt_nombre.Text
                    oUsuarioSelected.password = txt_password.Text
                    oUsuarioSelected.email = txt_email.Text
                    oUsuarioSelected.id_perfil = cbo_perfil.SelectedValue.ToString()

                    If oUsuarioService.actualizarUsuario(oUsuarioSelected) Then
                        MessageBox.Show("Usuario actualizado!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Me.Dispose()
                    Else
                        MessageBox.Show("Error al actualizar el usuario!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information)
                    End If
                End If

                Exit Select
            Case Opcion.delete

                If MessageBox.Show("Seguro que desea habilitar/deshabilitar el usuario seleccionado?", "Aviso", MessageBoxButtons.OKCancel, MessageBoxIcon.Question) = DialogResult.OK Then

                    If oUsuarioSelected.estado = "N" Then
                        oUsuarioSelected.estado = "S"
                    Else
                        oUsuarioSelected.estado = "N"
                    End If

                    If oUsuarioService.modificarEstadoUsuario(oUsuarioSelected) Then
                        MessageBox.Show("Usuario Habilitado/Deshabilitado!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information)
                        Me.Dispose()
                    Else
                        MessageBox.Show("Error al actualizar el usuario!", "Información", MessageBoxButtons.OK, MessageBoxIcon.Information)
                    End If
                End If

                Exit Select
        End Select
    End Sub

    Private Function validar_campos() As Boolean
        If txt_nombre.Text = String.Empty Then
            txt_nombre.BackColor = Color.Red
            txt_nombre.Focus()
            Return False
        Else
            txt_nombre.BackColor = Color.White
        End If

        If txt_password.Text = String.Empty Then
            txt_password.BackColor = Color.Red
            txt_password.Focus()
            Return False
        Else
            txt_password.BackColor = Color.White
        End If

        If txt_confirmar_pass.Text = String.Empty Then
            txt_confirmar_pass.BackColor = Color.Red
            txt_confirmar_pass.Focus()
            Return False
        Else
            txt_confirmar_pass.BackColor = Color.White
        End If

        If cbo_perfil.Text = String.Empty Then
            cbo_perfil.BackColor = Color.Red
            cbo_perfil.Focus()
            Return False
        Else
            cbo_perfil.BackColor = Color.White
        End If

        If txt_confirmar_pass.Text <> txt_password.Text Then
            txt_confirmar_pass.BackColor = Color.Red
            txt_password.BackColor = Color.Red
            txt_confirmar_pass.Focus()
            Return False
        Else
            txt_confirmar_pass.BackColor = Color.White
            txt_password.BackColor = Color.White
        End If

        Return True
    End Function

    Private Function existe_nombre() As Boolean
        Return (oUsuarioService.validarNombreUsuario(txt_nombre.Text) IsNot Nothing)
    End Function

    Private Sub llenarCombo(ByVal cbo As ComboBox, ByVal source As Object, ByVal display As String, ByVal value As String)
        cbo.DataSource = source
        cbo.DisplayMember = display
        cbo.ValueMember = value
        cbo.SelectedIndex = -1
    End Sub

    Private Sub btn_cancelar_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btn_cancelar.Click
        Me.Close()
    End Sub
End Class